JAICore/jaicore-ml/src/jaicore/ml/cache/ReproducibleInstances.java

```diff
@@ -0,0 +1,175 @@
+package jaicore.ml.cache;
+
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+import jaicore.ml.WekaUtil;
+import jaicore.ml.openml.OpenMLHelper;
+import weka.core.Instances;
+
+/**
+ * New Instances class to track splits and data origin.
+ * 
+ * @author jnowack
+ * @author mirko
+ * @author jonas
+ *
+ */
+public class ReproducibleInstances extends Instances {
+
+	/**
+	 * 
+	 */
+	private static final long serialVersionUID = 4318807226111536282L;
+	private List<Instruction> history = new LinkedList<>();
+	private boolean cacheStorage = true;
+	private boolean cacheLookup = true;
+	
+	private ReproducibleInstances(Instances dataset) {
+		super(dataset);
+	}
+	
+	/**
+	 * Creates a {@link ReproducibleInstances} Object based on the given History. Instructions that no not modify the Instances will be ignored (No evaluation will be done). 
+	 * 
+	 * @param history - List of INstructions used to create the original Instances
+	 * @param apiKey - apiKey in case openml.org is used
+	 * @return new {@link ReproducibleInstances} object
+	 * @throws IOException 
+	 * @throws  
+	 */
+	public static ReproducibleInstances fromHistory(List<Instruction> history, String apiKey) throws IOException {
+		ReproducibleInstances instances = null;
+		for (int i = 0; i < history.size(); i++) {
+			Instruction inst = history.get(i);
+			switch (inst.command) {
+			case "loadDataset":
```

Comment:
add the command ids as constants in the according command class and only reference them here via the constant. That's a good idea incase we want to change the ids later on for some reason.